### PR TITLE
redirect to requested page after sign-in

### PR DESCRIPTION
**Pull request by kylesferrazza**
closes https://github.com/CodeGrade/bottlenose/issues/112

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -1,4 +1,5 @@
 class UsersController < ApplicationController
+  before_action :require_current_user
```
[the remark this answers is not in the document]
Comment (kylesferrazza) [reply app/controllers/users_controller.rb]: Redirecting to `/users` means `require_site_admin` gets called, and the user gets redirected back to the root with "insufficient permission", not saving the `session[:next]`. Do you think this makes more sense:
```diff
diff --git a/app/controllers/users_controller.rb b/app/controllers/users_controller.rb
index 2bfcba7..f0fc95a 100644
--- a/app/controllers/users_controller.rb
+++ b/app/controllers/users_controller.rb
@@ -1,5 +1,4 @@
 class UsersController < ApplicationController
-  before_action :require_current_user
   before_action :require_site_admin, only: [:index]
   def index
     @users = User.order(:name)

```
```diff
diff --git a/app/controllers/application_controller.rb b/app/controllers/application_controller.rb
index ecf7e9d..de342fa 100644
--- a/app/controllers/application_controller.rb
+++ b/app/controllers/application_controller.rb
@@ -62,6 +62,7 @@ ERROR
 
   def require_site_admin
     unless current_user_site_admin?
+      session[:next] = request.fullpath
       redirect_to root_path, alert: "Must be an admin."
       return
     end
```

[the remark this answers is not in the document]
Comment (kylesferrazza) [reply app/controllers/users_controller.rb]: The problem with this is that a normal user trying to access `/users` before signing in will get indefinitely redirected after signing in.
What if we check for a logged in user in `require_site_admin`?

```diff
diff --git a/app/controllers/application_controller.rb b/app/controllers/application_controller.rb
index ecf7e9d..2707041 100644
--- a/app/controllers/application_controller.rb
+++ b/app/controllers/application_controller.rb
@@ -61,6 +61,11 @@ ERROR
   end
 
   def require_site_admin
+    if current_user.nil?
+      session[:next] = request.fullpath
+      redirect_to root_path, alert: "You need to log in first."
+      return
+    end
     unless current_user_site_admin?
       redirect_to root_path, alert: "Must be an admin."
       return
```
```diff
diff --git a/app/controllers/users_controller.rb b/app/controllers/users_controller.rb
index 2bfcba7..f0fc95a 100644
--- a/app/controllers/users_controller.rb
+++ b/app/controllers/users_controller.rb
@@ -1,5 +1,4 @@
 class UsersController < ApplicationController
-  before_action :require_current_user
   before_action :require_site_admin, only: [:index]
   def index
     @users = User.order(:name)
```

[the remark this answers is not in the document]
Comment (blerner) [reply app/controllers/users_controller.rb]: That seems more reasonable, yes.

[the remark this answers is not in the document]
Comment (kylesferrazza) [reply app/controllers/users_controller.rb]: Ok, updated.

**Comment by kylesferrazza** [comment]
Need anything else on this one?

**Comment by blerner** [comment]
Looks good. I'll try it out later if I have time today, and then merge.

**Comment by blerner** [comment]
Mmm, yes, that would be a welcome improvement!

**Comment by kylesferrazza** [comment]
I've been trying to figure out how to pass a param through Devise and have it come out the other side...

This seems useful, but I haven't been able to figure it out yet: https://github.com/plataformatec/devise/wiki/How-To:-redirect-to-a-specific-page-on-successful-sign-in

**Pull request merged**